diesel/src/macros/insertable.rs

```diff
@@ -0,0 +1,422 @@
+/// Implements the [`Insertable`][insertable] trait for a given struct. This
+/// macro should be called with the name of the table you wish to use the struct
+/// with, followed by the entire struct body.
+///
+/// [insertable]: prelude/trait.Insertable.html
+///
+/// # Example
+///
+/// ```no_run
+/// # #[macro_use] extern crate diesel;
+/// # table! { users { id -> Integer, name -> VarChar, hair_color -> Nullable<VarChar>, } }
+/// struct NewUser<'a> {
+///     name: &'a str,
+///     hair_color: &'a str,
+/// }
+///
+/// Insertable! {
+///     (users)
+///     struct NewUser<'a> {
+///         name: &'a str,
+///         hair_color: &'a str,
+///     }
+/// }
+/// # fn main() {}
+/// ```
+///
+/// To avoid copying your struct definition, you can use the
+/// [custom_derive crate][custom_derive].
+///
+/// ```ignore
+/// custom_derive! {
+///     #[derive(Insertable(users))]
+///     struct NewUser<'a> {
+///         name: &'a str,
+///         hair_color: &'a str,
+///     }
+/// }
+/// ```
+///
+/// You can also use this macro with tuple structs, but *all* fields must be
+/// annotated with `#[column_name(name)]`. Additionally, a trailing comma after
+/// the last field is required.
+///
+/// ```no_run
+/// # #[macro_use] extern crate diesel;
+/// # table! { users { id -> Integer, name -> VarChar, hair_color -> Nullable<VarChar>, } }
+/// struct NewUser<'a>(&'a str, Option<&'a str>);
+///
+/// Insertable! {
+///     (users)
+///     struct NewUser<'a>(
+///         #[column_name(name)]
+///         &'a str,
+///         #[column_name(hair_color)]
+///         Option<&'a str>,
+///     );
+/// }
+/// # fn main() {}
+/// ```
+#[macro_export]
+macro_rules! Insertable {
+    // Strip meta items, pub (if present) and struct from definition
+    (
+        ($table_name:ident)
+        $(#[$ignore:meta])*
+        $(pub)* struct $($body:tt)*
+    ) => {
+        Insertable! {
+            ($table_name)
+            $($body)*
+        }
+    };
+
+    // Handle struct with lifetimes
```

Comment:
Yup. Would normally restructure this code to not need it, but that's basically impossible with this type of macro.